0000-interledger-protocol-4/0000-interledger-protocol-4.md

```diff
@@ -0,0 +1,159 @@
+10. Sender checks that the preimage in the ILP Fulfill packet matches the condition in their Prepare packet and may read any data returned by the receiver.
+11. Sender signs an update to their payment channel with the first connector to cover the value of the ILP Prepare (or they may settle less frequently using slower on-ledger transfers or physical deliveries of cash or other goods) and communicates it to the connector.
+12. Connector validates the sender's signed payment channel update (or other proof of settlement) and increases the sender's balance by the amount they paid. If the sender does not pay for the ILP Prepare, the connector may refuse to process further ILP packets until they do or the connector may block that user entirely.
+
+## Specification
+
+### Packet Format
+
+ILPv4 packets SHOULD be encoded using the ASN.1 Octet Encoding Rules.
+
+#### Type-Length Wrapper
+
+All ILP packets are wrapped in the following envelope:
+
+| Field | Type | Description |
+|---|---|---|
+| `type` | UInt8 | ID of the ILP Packet type |
+| `data` | Variable-Length Octet String | Packet contents, prefixed with their length |
+
+See the ASN.1 definition [here](../asn1/InterledgerPacket.asn).
+
+#### ILP Prepare
+
+ILP Prepare packets are `type` 12.
+
+| Field | Type | Description |
+|---|---|---|
+| `amount` | UInt64 | Local amount, denominated in the minimum divisible unit of the asset of the bilateral relationship. This field is modified by each connector, who applies their exchange rate before forwarding the packet |
+| `expiresAt` | Fixed-Length [Interledger Timestamp](../asn1/InterledgerTypes.asn) | Date and time when the packet expires. Each connector changes the value of this field to set the expiry to an earlier time, before forwarding the packet. | 
+| `executionCondition` | UInt256 | SHA-256 hash digest of the `fulfillment` that will execute the transfer of value represented by this packet |
+| `destination` | [ILP Address](../0015-ilp-addresses/0015-ilp-addresses.md) | ILP Address of the receiver |
+| `data` | Variable-Length Octet String | End-to-end data. Connectors MUST NOT modify this data. Most higher-level protocols will encrypt and authenticate this data, so receivers will reject packets in which the data is modified | 
+
+**Note:** There is no `from` or `source` address in the ILP Prepare packet. Each hop MUST determine which of their immediate peers or customers the packet came from (to debit the correct account) based on the communication channel the packet was received through. Also, higher-level protocols MAY communicate the sender's address to the receiver if the use case calls for receivers to send ILP Prepare packets to the sender (they can communicate responses to the sender using the `data` field in ILP Fulfill and Reject packets).
```

Comment:
Based on communication channel and authentication information (e.g. in the case of an HTTP endpoint)

Reply:
@sharafian depending on your definition of "communication channel" that is implied.

http://www.enterpriseintegrationpatterns.com/patterns/messaging/PointToPointChannel.html